subprojects/core/src/main/java/org/gradle/process/internal/DefaultJavaForkOptions.java

```diff
@@ -148,4 +158,131 @@ public JavaForkOptions copyTo(JavaForkOptions target) {
+    }
+
+    @Override
+    public boolean isCompatibleWith(JavaForkOptions options) {
+        return getDebug() == options.getDebug()
+                && getEnableAssertions() == options.getEnableAssertions()
+                && normalized(getExecutable()).equals(normalized(options.getExecutable()))
+                && getWorkingDir().equals(options.getWorkingDir())
+                && normalized(getDefaultCharacterEncoding()).equals(normalized(options.getDefaultCharacterEncoding()))
+                && getHeapSizeMb(getMinHeapSize()) >= getHeapSizeMb(options.getMinHeapSize())
+                && getHeapSizeMb(getMaxHeapSize()) >= getHeapSizeMb(options.getMaxHeapSize())
+                && normalized(getJvmArgs()).containsAll(normalized(options.getJvmArgs()))
+                && containsAll(getSystemProperties(), options.getSystemProperties())
+                && containsAll(getEnvironment(), options.getEnvironment())
+                && getBootstrapClasspath().getFiles().containsAll(options.getBootstrapClasspath().getFiles());
+    }
+
+    private int getHeapSizeMb(String heapSize) {
+        if (heapSize == null) {
+            return -1; // unspecified
+        }
+
+        String normalized = heapSize.trim().toLowerCase();
+        try {
+            if (normalized.endsWith("m")) {
+                return Integer.parseInt(normalized.substring(0, normalized.length() - 1));
+            }
+            if (normalized.endsWith("g")) {
+                return Integer.parseInt(normalized.substring(0, normalized.length() - 1)) * 1024;
+            }
+        } catch (NumberFormatException e) {
+            throw new InvalidUserDataException("Cannot parse heap size: " + heapSize, e);
+        }
+        throw new InvalidUserDataException("Cannot parse heap size: " + heapSize);
+    }
+
+    private String mergeHeapSize(String heapSize1, String heapSize2) {
+        int mergedHeapSizeMb = Math.max(getHeapSizeMb(heapSize1), getHeapSizeMb(heapSize2));
+        return mergedHeapSizeMb == -1 ? null : String.valueOf(mergedHeapSizeMb) + "m";
+    }
+
+    private boolean canBeMerged(String left, String right) {
+        if (left == null || right == null) {
+            return true;
+        } else {
+            return normalized(left).equals(normalized(right));
+        }
+    }
+
+    private boolean canBeMerged(File left, File right) {
+        if (left == null || right == null) {
+            return true;
+        } else {
+            return left.equals(right);
+        }
+    }
+
+    private Set<String> normalized(Iterable<String> strings) {
+        Set<String> normalized = Sets.newLinkedHashSet();
+        for (String string : strings) {
+            normalized.add(normalized(string));
+        }
+        return normalized;
+    }
+
+    private String normalized(String string) {
+        return nullToEmpty(string).trim();
+    }
+
+    private boolean containsAll(Map<String, Object> left, Map<String, Object> right) {
```

Comment:
I don't think it's as simple as that - the keys would have to be exactly the same for that to pass (instead of allowing for normalized equality of keys).  Of course, we could have a method to normalize the entrySet.  Or we could just remove the normalization and say that "foo   " != "foo", but I was simply duplicating the logic we had in `DaemonForkOptions` for checking compatibility, so that would be a behavior change.